internal/jobcontainers/path.go

```diff
@@ -0,0 +1,143 @@
+package jobcontainers
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/windows"
+)
+
+func searchPathForExe(fileName, pathsToSearch string) (string, error) {
+	fileNamePtr, err := windows.UTF16PtrFromString(fileName)
+	if err != nil {
+		return "", err
+	}
+
+	pathsToSearchPtr, err := windows.UTF16PtrFromString(pathsToSearch)
+	if err != nil {
+		return "", err
+	}
+
+	extension, err := windows.UTF16PtrFromString(".exe")
+	if err != nil {
+		return "", err
+	}
+
+	path := make([]uint16, windows.MAX_PATH)
+	_, err = winapi.SearchPath(
+		pathsToSearchPtr,
+		fileNamePtr,
+		extension,
+		windows.MAX_PATH,
+		&path[0],
+		nil,
+	)
+	if err != nil {
+		return "", err
+	}
+
+	exePath := windows.UTF16PtrToString(&path[0])
+	// Need to check if we just found a directory with the name of the executable and
+	// .exe at the end. ping.exe is a perfectly valid directory name for example.
+	attrs, err := os.Stat(exePath)
+	if err != nil {
+		return "", &os.PathError{Path: exePath, Op: "stat", Err: err}
+	}
+
+	if attrs.IsDir() {
+		return "", fmt.Errorf("found directory instead of executable %q", exePath)
+	}
+
+	return exePath, nil
+}
+
+// Returns the system paths (system32, system, and windows) as a search path,
+// including a terminating ;.
+//
+// Typical output would be `C:\WINDOWS\system32;C:\WINDOWS\System;C:\WINDOWS;`
+func getSystemPaths() (string, error) {
+	var searchPath string
+	systemDir, err := windows.GetSystemDirectory()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get system directory")
+	}
+	searchPath += systemDir + ";"
+
+	windowsDir, err := windows.GetWindowsDirectory()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get Windows directory")
+	}
+
+	searchPath += windowsDir + "\\System;" + windowsDir + ";"
+	return searchPath, nil
+}
+
+// Gets the application name from the commandline provided and returns the commandline
+// as is or quoted if found necessary.
+func getApplicationName(commandLine, workingDirectory, pathEnv string) (string, string, error) {
```

Comment:
Since we end up adjusting the command line at the end anyways, why not just immediately adjust the command line at the beginning of this function and then directly search for it. In that scenario, a command line of `hello goodby.exe` would be treated as `"hello goodbye.exe"` which I feel like is a reasonable thing to expect.

Reply:
I don't understand but I might not be following your question. We'd only know whether it needs to be adjusted at the end after we've searched piece by piece and one of them ends up succeeding.

Reply:
Take this for example:
`foo bar baz` 
If `foo.exe` is successfully found we can stop right there and return with the full path to foo. If foo doesn't succeed we then try `foo bar.exe` and so on. If `foo bar.exe` is found (or `foo bar baz.exe`, then we need to quote.

Reply:
I think it's pretty common to enforce that any path or exe that has a space in it would need to be quoted. I feel like that's reasonable to enforce here.

Reply:
I think what's missing here is a big comment explaining what the resolution logic is, some examples, and most importantly why it's written this way (namely, that it matches the existing behavior of CreateProcess and Windows silo containers).

Reply:
This is a good example of why it's really important to document the "why" of logic in code.

Reply:
I don't think we should change the logic people are used to for container commands so I don't agree that we should require it to be quoted. I DO agree that I should document this more 😆

Reply:
Would y'all rather have comments above each logical step, or one large comment above getApplicationName?